Bluetooth: ISO: Fixes and improvements

Vudentz opened the pull request:
A couples more fixes for properly handling CIS request and enable bt connections to list ISO connections.

Vudentz committed:
Bluetooth: shell: Add connection type information to connections command

This prints information about the role and id along with connection
parameters in case the connection is of LE type.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: conn: Add support for ISO connections to bt_conn_get_info

This enables bt_conn_get_info to be used with connection of type
BT_CONN_TYPE_ISO.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ISO: Simplify reference handling

This makes bt_conn_cleanup to be automatically called when the last
reference to a connection of BT_CONN_TYPE_ISO is dropped.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ISO: Fix not assigning CIG and CIS IDs

When handling CIS Requested event CIG and CIS IDs must be set in order
for the channel to be properly accepted.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ISO: Fix call order of connected callback

The state shall be set before calling the callback as it could be check
code behind would expect the channel to be in connected state.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan reviewed (dismissed):
Unsafe to still use the conn object after the reference count is 0.

Vudentz commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1538,7 +1538,9 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 		 * again.
 		 */
-		bt_conn_ref(conn);
+		if (conn->type == BT_CONN_TYPE_LE) {
```
Well the comment talks about bt_conn_add_le so I suppose this logic is only valid for LE or does BR rely on it as well?

Vudentz commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1894,10 +1895,15 @@ void bt_conn_unref(struct bt_conn *conn)
+
+	if (IS_ENABLED(CONFIG_BT_ISO) && conn->type == BT_CONN_TYPE_ISO &&
+	    atomic_get(&conn->ref) == 0) {
+		bt_iso_cleanup(conn);
```
Right, I will change the logic to only drop the reference after cleanup.

Vudentz commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1942,6 +1942,12 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
 		return 0;
+#endif
+#if defined(CONFIG_BT_ISO)
+	case BT_CONN_TYPE_ISO:
+		info->le.dst = &conn->iso.acl->le.dst;
+		info->le.src = &bt_dev.id_addr[conn->iso.acl->id];
+		return 0;
```
Will do, I guess for broadcast we don't actually have anything to fill those addresses, or do we?

Vudentz replied:
Ive added a check for bis for now, we can latter add more information to bt_conn_info for CIS/BIS regarding the QoS, etc.

Vudentz commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -2396,13 +2396,52 @@ static int cmd_bonds(const struct shell *shell, size_t argc, char *argv[])
+			    info.le.interval, info.le.latency, info.le.timeout);
+		break;
+#if defined(CONFIG_BT_ISO)
+	case BT_CONN_TYPE_ISO:
+		bt_addr_le_to_str(info.le.dst, addr, sizeof(addr));
+		shell_print(ctx_shell, "#%u [ISO][%s] %s", info.id,
+			    role_str(info.role), addr);
+		break;
```
For now it will just print FF:..:FF as address, latter on we can add QoS parameters, etc.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1887,14 +1888,21 @@ struct bt_conn *bt_conn_ref(struct bt_conn *conn)
 void bt_conn_unref(struct bt_conn *conn)
 {
-	atomic_val_t old = atomic_dec(&conn->ref);
+	atomic_val_t old;
+
+	if (IS_ENABLED(CONFIG_BT_ISO) && conn->type == BT_CONN_TYPE_ISO &&
+	    atomic_get(&conn->ref) == 1) {
+		bt_iso_cleanup(conn);
+	}
+
+	old = atomic_dec(&conn->ref);
```
I assume the reason for this change, is that we need to call `bt_iso_cleanup` with the ref "intact" for some reason. Could you possibly add a comment stating why we need this special handling for ISO channels?

Vudentz replied:
Yep, @joerchan pointed out that if we release the reference another thread may, in the meantime, allocated while bt_iso_cleanup has not completed it.

Vudentz replied:
Done.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1936,6 +1944,17 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+#if defined(CONFIG_BT_ISO)
+	case BT_CONN_TYPE_ISO:
+		if (!conn->iso.is_bis) {
+			info->le.dst = &conn->iso.acl->le.dst;
+			info->le.src = &bt_dev.id_addr[conn->iso.acl->id];
```
Can this function be called on a ISO bt_conn where the ACL isn't established? If so, we need a check for `conn->acl`.

Vudentz replied:
CIS always shall be bound to an ACL otherwise it cannot exist.

Vudentz commented:
@Thalley @joerchan I wonder if you guys can review it as addressed your comments.

Thalley reviewed (approved):
Assuming that it has been tested to work, LGTM

Vudentz committed:
Bluetooth: conn: Align cleanup for ISO connections

This makes connections of type BT_CONN_TYPE_ISO use BT_CONN_CLEANUP to
cleanup so the TX and RX queues are properly cleanup.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented on subsys/bluetooth/host/conn.c:
```diff
@@ -346,6 +346,35 @@ static void tx_complete_work(struct k_work *work)
+		bt_conn_unref(iso_conn);
+	}
+
+	return NULL;
+#else
+	return NULL;
+#endif /* CONFIG_BT_ISO */
```
Yep, I will fix it.

Thalley commented:
@Vudentz There's something in regards to the ref counter that doesn't quite work with this PR. 

A test where the peripheral disconnect, not only causes the CIS conn not to reach ref counter 0, but makes the ACL conn underflow. 

See this following log from a disconnect: 
```
[00:00:01.230,000] <dbg> bt_conn.bt_conn_unref: handle 2 ref 2 -> 1
[00:00:01.230,000] <dbg> bt_iso.bt_iso_chan_send: chan 0x5659da10 len 1
[00:00:01.230,000] <dbg> bt_conn.bt_conn_send_cb: conn handle 2 buf len 5 cb (nil) user_data (nil)
[00:00:01.230,000] <dbg> bt_conn.bt_conn_process_tx: conn 0x5659d740
[00:00:01.230,000] <dbg> bt_conn.send_buf: conn 0x5659d740 buf 0x5659dc9c len 5
[00:00:01.230,000] <dbg> bt_conn.send_frag: conn 0x5659d740 buf 0x5659dc9c len 5 flags 0x02
[00:00:01.230,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:01.230,000] <dbg> bt_conn.conn_prepare_events: Adding conn 0x5659d460 to poll list
[00:00:01.230,000] <dbg> bt_conn.conn_prepare_events: Adding conn 0x5659d740 to poll list
[00:00:01.290,000] <dbg> bt_conn.bt_conn_ref: handle 0 ref 3 -> 4
[00:00:01.290,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 4 -> 3
[00:00:02.240,000] <dbg> bt_iso.bt_iso_chan_send: chan 0x5659da10 len 2
[00:00:02.240,000] <dbg> bt_conn.bt_conn_send_cb: conn handle 2 buf len 6 cb (nil) user_data (nil)
[00:00:02.240,000] <dbg> bt_conn.bt_conn_process_tx: conn 0x5659d740
[00:00:02.240,000] <dbg> bt_conn.send_buf: conn 0x5659d740 buf 0x5659dc9c len 6
[00:00:02.240,000] <dbg> bt_conn.send_frag: conn 0x5659d740 buf 0x5659dc9c len 6 flags 0x02
[00:00:05.910,000] <dbg> bt_conn.bt_conn_ref: handle 0 ref 3 -> 4
[00:00:05.910,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 4 -> 3
[00:00:49.890,000] <dbg> bt_conn.bt_conn_ref: handle 0 ref 3 -> 4
[00:00:49.890,000] <dbg> bt_conn.bt_conn_set_state: connected -> disconnect-complete
[00:00:49.890,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 4 -> 3
[00:00:49.890,000] <dbg> bt_conn.bt_conn_ref: handle 0 ref 3 -> 4
[00:00:49.890,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 4 -> 3
[00:00:49.890,000] <dbg> bt_conn.bt_conn_ref: handle 2 ref 1 -> 2
[00:00:49.890,000] <dbg> bt_conn.bt_conn_set_state: connected -> disconnect-complete
[00:00:49.890,000] <dbg> bt_conn.bt_conn_unref: handle 2 ref 2 -> 1
[00:00:49.890,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:49.890,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:49.890,000] <dbg> bt_conn.bt_conn_ref: handle 0 ref 3 -> 4
[00:00:49.890,000] <dbg> bt_conn.bt_conn_set_state: disconnect-complete -> disconnected
[00:00:49.890,000] <dbg> bt_conn.tx_notify: conn 0x5659d460
[00:00:49.890,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 4 -> 3
[00:00:49.890,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:49.890,000] <dbg> bt_conn.bt_conn_ref: handle 0 ref 3 -> 4
[00:00:49.890,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 4 -> 3
**[00:00:49.890,000] <dbg> bt_conn.bt_conn_ref: handle 2 ref 1 -> 2**
[00:00:49.890,000] <dbg> bt_conn.bt_conn_set_state: disconnect-complete -> disconnected
[00:00:49.890,000] <dbg> bt_conn.tx_notify: conn 0x5659d740
[00:00:49.890,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 2 -> 1
[00:00:49.890,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:49.890,000] <dbg> bt_iso.bt_iso_chan_send: chan 0x5659da10 len 3
[00:00:49.890,000] <dbg> bt_conn.bt_conn_send_cb: conn handle 0 buf len 7 cb (nil) user_data (nil)
[00:00:49.890,000] <err> bt_conn: not connected!
Failed to send ISO data (-107)
[00:00:49.890,000] <dbg> bt_conn.deferred_work: conn 0x5659d460
[00:00:49.890,000] <dbg> bt_conn.bt_conn_ref: handle 0 ref 1 -> 2
[00:00:49.890,000] <dbg> bt_iso.bt_iso_disconnected: 0x5659d740
[00:00:49.890,000] <dbg> bt_iso.bt_iso_remove_data_path: 0x5659d740
[00:00:49.920,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:49.950,000] <wrn> bt_hci_core: opcode 0x206f status 0x02
[00:00:49.950,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:49.980,000] <wrn> bt_hci_core: opcode 0x206f status 0x02
[00:00:49.980,000] <dbg> bt_iso.bt_iso_chan_disconnected: 0x5659da10, reason 0x08
[00:00:49.980,000] <dbg> bt_iso.bt_iso_chan_set_state_debug: chan 0x5659da10 conn 0x5659d740 connected -> bound
[00:00:49.980,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 2 -> 1
[00:00:49.980,000] <dbg> bt_iso.bt_iso_chan_set_state_debug: chan 0x5659da10 conn (nil) bound -> disconnected
ISO Channel 0x5659da10 disconnected (reason 0x08)
[00:00:49.980,000] <dbg> bt_iso.bt_iso_cleanup: 0x5659d740
[00:00:49.980,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 3 -> 2
[00:00:49.980,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:50.010,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 1 -> 0
Disconnected: FA:4C:4B:DB:D3:89 (public) (reason 0x08)
[00:00:50.010,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 2 -> 1
[00:00:50.010,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:50.040,000] <dbg> bt_conn.bt_conn_prepare_events: 
[00:00:50.070,000] <dbg> bt_conn.bt_conn_prepare_events: 
Scanning successfully started
[00:00:50.100,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 1 -> 0
[00:00:50.100,000] <dbg> bt_conn.deferred_work: conn 0x5659d740
**[00:00:50.100,000] <dbg> bt_conn.bt_conn_unref: handle 0 ref 0 -> 4294967295**
```

I've tried to figure out where the issue is, but I can't pinpoint it. 

The issue can easily be reproduced by running the peripheral_iso and central_iso samples, and simply disconnect the peripheral in some way.

Thalley reviewed (changes requested):
See https://github.com/zephyrproject-rtos/zephyr/pull/33889#issuecomment-831912292

Vudentz commented:
> See [#33889 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/33889#issuecomment-831912292)

Patch has been updated.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -339,6 +339,32 @@ static void tx_complete_work(struct k_work *work)
 	tx_notify(conn);
 }
 
+struct bt_conn *conn_lookup_iso(struct bt_conn *conn)
```
Shouldn’t this be static?

Vudentz replied:
Fixes.
